### DISCUSSION TITLE
Use with Apollo Server Cache

### DISCUSSION THREAD

**Discussion by tylerjbainbridge** [hayes/pothos#1136] [General]
Hi @hayes -- first off. Thank you for building this framework, it's incredible.

I'm really struggling to understand how one would go about integrating server level caching with Pothos builders & resolvers.

Would you be able to help here and perhaps provide a concise example or two of how I'd go about caching certain fields using `info.cacheControl.setCacheHint`?

**Comment by hayes** [comment]
Are you running into specific issues?  the `info` object is the 4th argument of any resolver (5th for t.prismaField).  Pothos doesn't really interact with this directly, the resolvers are just added to the schema the same way they would be for any other way of defining resolvers, so it should work the same as anything documented by apollo

**Reply by tylerjbainbridge** [reply]
I see the fields on info that are passed down via Apollo Server, but ideally I can pass directives in to the fields on to types themselves like so-

```
user: t.relation('user', {
      nullable: true,
      directives: {
        cacheControl: {
          maxAge: 1000,
        },
      },
    }),
```

I saw your reply on my issue, but does the directives plugin need to be imported in order to do something like this?

**Reply by hayes** [reply]
yes, the directive plugin is required to define directives like this

**Reply by hayes** [reply]
I think you've already found this, but the comment here: https://github.com/hayes/pothos/issues/1059#issuecomment-1749047071 describes the best way to set this up, if we can figure out why the plugin isn't working correctly for you

**Comment by tylerjbainbridge** [comment]
@hayes I'm looking for some clarity on how Pothos actually works, because I believe it takes the Graphql Query itself and uses includes/selects to compile one query that Prisma sends off to the database.

So if this is the case, is it impossible to cache a specific type/id? because it'd be too late-- the query details have already been included in the top level prisma call-- so even if you have a certain user or post cached already, it will be fetched from the database either way. 

Maybe this is showing a fundamental misunderstanding of how Pothos and/or GraphQL queries are processed, but I'd love to hear your thoughts.

**Reply by hayes** [reply]
So apollo caching and the cache control directives don't actually cache things at a type/field level.  What it actually does is finds the lowest maxAge for the fields in a query which it uses to configure a cache-control header on the response.  From there, it works like normal http caching where the full response is cached/invalidated together.  The actual caching can happen at a bunch of different layers (client, browser, cdn, proxy, gateway, etc) depending on how your request is routed, and how any intermediate services are configured.

Your understanding of the Prisma plugin for pothos is roughly correct, it will load as much data for the request as possible in the root resolver, so most nested fields/types won't need to make additional db queries.  Pothos can't always load everything in the root resolver though, and some things will still be differed to the nested resolvers.   This generally means that the prisma plugin doesn't work well with things like data-loaders, or caching built into the GraphQL API itself.  But the way that apollos cache-control directives work should work exactly the same with pothos as any other GraphQL API.
